### PR TITLE
Add PHPFPM stat

### PR DESCRIPTION
**Pull request by kureikain**
- HTTP status or Socket status
- Collect those metric:
    accepted conn:
    listen queue:
    max listen queue:
    listen queue len:
    idle processes:
    active processes:
    total processes:
    max active processes:
    max children reached:
    slow requests:
- Tag metric with: `host` and `pool` name

### CODE REVIEW

**Comment by sparrc** [comment]
Thank you @kureikain! It looks like `godep go vet ./...` has failed:

```
Executing 'godep go vet ./...'
plugins/phpfpm/phpfpm_fcgi.go:131: net.UnixAddr composite literal uses unkeyed fields
exit status 1
godep: go exit status 1
```

**Comment by sparrc** [comment]
@kureikain can you explain why the fast cgi client is necessary for this? It looks like a rather simple and small http status page so I'm not sure I see the justification for all of that.

**Comment by kureikain** [comment]
@sparrc The status can also be fetched using socket via FastCGI protocol. On Ubuntu, people use this `http://manpages.ubuntu.com/manpages/utopic/man1/cgi-fcgi.1.html` to fetch status using socket.

This user pointed out they want it https://github.com/influxdb/telegraf/issues/234#issuecomment-146416055

So if user want to use http status page like:

```
http://ip/status
```

We used the HTTP client 

If they specified:

```
192.168.10.10:/var/run/php5-fpm.sock
```

The socket is used, and we need that fastcgi client. https://github.com/influxdb/telegraf/pull/255/files#diff-8c5d9b738b687989ac2793e76195cd00R53

File: plugins/phpfpm/phpfpm.go [hunk 1]
```diff
@@ -0,0 +1,187 @@
...
+
+// Request status page to get stat raw data
+func (g *phpfpm) gatherServer(addr string, acc plugins.Accumulator) error {
+	if g.client == nil {
+
+		client := &http.Client{}
+		g.client = client
+	}
+
+	if strings.HasPrefix(addr, "http://") || strings.HasPrefix(addr, "https://") {
```
Comment (kureikain) [inline plugins/phpfpm/phpfpm.go]: @sparrc Here, we are requesting via HTTP

**Comment by sparrc** [comment]
I see, thanks for the clarification, I have a few more comments but otherwise looks good :+1:

File: plugins/phpfpm/phpfpm.go [hunk 2]
```diff
@@ -0,0 +1,187 @@
...
+
+		if res.StatusCode != 200 {
+			return fmt.Errorf("Unable to get valid stat result from '%s': %v", addr, err)
+		}
+
+		importMetric(res.Body, acc, u.Host)
+	} else {
+		socketAddr := strings.Split(addr, ":")
+
+		fcgi, _ := NewClient("unix", socketAddr[1])
```
Comment (kureikain) [inline plugins/phpfpm/phpfpm.go]: Here, we are requesting via FastCGI, using the socket liket `/var/run/php5-fpm.sock`

File: plugins/phpfpm/phpfpm.go [hunk 3]
```diff
@@ -0,0 +1,187 @@
...
+var sampleConfig = `
+	# An array of address to gather stats about. Specify an ip on hostname
+	# with optional port and path. ie localhost, 10.10.3.33/server-status, etc.
+	#
+	# We can configure int two modes:
+	#   - unixsocket: the string is the path to fpm socket like /var/run/php5-fpm.sock
+	#   - http: the URL has to start with http:// or https://
+	#
+	# If no servers are specified, then default to 127.0.0.1/server-status
+	statusurls = ["http://localhost/status"]
```
Comment (sparrc) [inline plugins/phpfpm/phpfpm.go]: rename `statusurls` to `urls` for consistency

File: plugins/phpfpm/phpfpm.go [hunk 4]
```diff
@@ -0,0 +1,187 @@
...
+	# statusurls = ["http://localhost/status?full"]
+	#
+	# Or you can also use local socket.
+	# statusurls = ["192.168.10.10:/var/run/php5-fpm-www2.sock", "192.168.10.10:/var/run/php5-fpm-www1.sock"]
```
Comment (sparrc) [inline plugins/phpfpm/phpfpm.go]: shorten these lines to 80 characters

File: plugins/phpfpm/phpfpm.go [hunk 5]
```diff
@@ -0,0 +1,187 @@
...
+		// We start to gather data for a new pool here
+		if fieldName == PF_POOL {
+			currentPool = strings.Trim(keyvalue[1], " ")
+			stats[currentPool] = make(metric)
+			continue
+		}
+
+		// Start to parse metric for current pool
+		switch fieldName {
+		case PF_ACCEPTED_CONN, PF_LISTEN_QUEUE, PF_MAX_LISTEN_QUEUE, PF_LISTEN_QUEUE_LEN, PF_IDLE_PROCESSES, PF_ACTIVE_PROCESSES, PF_TOTAL_PROCESSES, PF_MAX_ACTIVE_PROCESSES, PF_MAX_CHILDREN_REACHED, PF_SLOW_REQUESTS:
```
Comment (sparrc) [inline plugins/phpfpm/phpfpm.go]: line length

File: plugins/phpfpm/phpfpm_fcgi.go [hunk 6]
```diff
@@ -0,0 +1,321 @@
...
+		err = errors.New("fcgi: we only accept int (port) or string (socket) params.")
+	}
+	fcgi = &FCGIClient{
+		rwc:       conn,
+		keepAlive: false,
+	}
+	return
+}
+
+func (this *FCGIClient) writeRecord(recType uint8, reqId uint16, content []byte) (err error) {
```
Comment (sparrc) [inline plugins/phpfpm/phpfpm_fcgi.go]: Please change the receiver name, using `this` or `self` is strongly discouraged in Go, see https://github.com/golang/go/wiki/CodeReviewComments#receiver-names

**Comment by sparrc** [comment]
@kureikain you'll also need to expand on the README, use https://github.com/influxdb/telegraf/blob/master/plugins/system/CPU_README.md as an example. The main thing this is missing is a list of the metrics it will output.

**Comment by kureikain** [comment]
@sparrc Ok, I fixed code per review :). Update README. Hopefully the test will passed :)

**Comment by sparrc** [comment]
Thank you @kureikain! One more thing: Telegraf uses `host` as a default tag for the host that the telegraf instance is running on, you are overriding this with your `host` tag.

Could you please change your tag to `url`?

**Comment by kureikain** [comment]
@sparrc My bad :(. Just updated it as your suggestion

**Comment by sparrc** [comment]
no problem, thank you @kureikain! I'm going to merge this now

**Comment by kokarn** [comment]
It's missing from the main readme tho :) 
Should I do a pull or can you fix it?

**Comment by kureikain** [comment]
@kokarn Can you did it and send the PR?

**Comment by kokarn** [comment]
Sure, super easy. #274

**Pull request closed without merging**